<a href="https://colab.research.google.com/github/Deepanshuduttpant/wallmart-sale/blob/main/Wallmart_sales_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd 
import numpy as np
import plotly.express as px

In [ ]:
 df1 = pd.read_csv('/content/Walmart.csv')

# Basic level analysis

In [ ]:
df1.head() 

Order ID  Order Date   Ship Date    Customer Name        Country  \
0  CA-2013-138688  13-06-2013  17-06-2013  Darrin Van Huff  United States   
1  CA-2011-115812  09-06-2011  14-06-2011  Brosina Hoffman  United States   
2  CA-2011-115812  09-06-2011  14-06-2011  Brosina Hoffman  United States   
3  CA-2011-115812  09-06-2011  14-06-2011  Brosina Hoffman  United States   
4  CA-2011-115812  09-06-2011  14-06-2011  Brosina Hoffman  United States   

          City       State     Category  \
0  Los Angeles  California       Labels   
1  Los Angeles  California  Furnishings   
2  Los Angeles  California          Art   
3  Los Angeles  California       Phones   
4  Los Angeles  California      Binders   

                                        Product Name   Sales  Quantity  Profit  
0  Self-Adhesive Address Labels for Typewriters b...   14.62       2.0    6.87  
1  Eldon Expressions Wood and Plastic Desk Access...   48.86       7.0   14.17  
2                                         Newell 322    7.28       4.0    1.97  
3                     Mitel 5320 IP Phone VoIP phone  907.15       4.0   90.72  
4  DXL Angle-View Binders with Locking Rings by S...   18.50       3.0    5.78

to see the total no of rows and columns

In [ ]:
df1.shape

(3203, 12)

to see the names of the columns

In [ ]:
list(df1.columns)

['Order ID',
 'Order Date',
 'Ship Date',
 'Customer Name',
 'Country',
 'City',
 'State',
 'Category',
 'Product Name',
 'Sales',
 'Quantity',
 'Profit']

In [ ]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3203 entries, 0 to 3202
Data columns (total 12 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Order ID       3203 non-null   object 
 1   Order Date     3203 non-null   object 
 2   Ship Date      3203 non-null   object 
 3   Customer Name  3203 non-null   object 
 4   Country        3203 non-null   object 
 5   City           3203 non-null   object 
 6   State          3203 non-null   object 
 7   Category       3203 non-null   object 
 8   Product Name   3203 non-null   object 
 9   Sales          3203 non-null   float64
 10  Quantity       3203 non-null   float64
 11  Profit         3203 non-null   float64
dtypes: float64(3), object(9)
memory usage: 300.4+ KB


to see the null values 

In [ ]:
df1.isna().sum()

Order ID         0
Order Date       0
Ship Date        0
Customer Name    0
Country          0
City             0
State            0
Category         0
Product Name     0
Sales            0
Quantity         0
Profit           0
dtype: int64

In [ ]:
df1.nunique()

Order ID         1611
Order Date        845
Ship Date         911
Customer Name     686
Country             1
City              169
State              11
Category           17
Product Name     1494
Sales            2295
Quantity           14
Profit           2164
dtype: int64

converting order date and ship date datatype to datetime 

In [ ]:
df1['Order Date'] = pd.to_datetime(df1['Order Date'])
df1['Ship Date'] = pd.to_datetime(df1['Ship Date'])

<ipython-input-118-a7fea54ccc0d>:1: UserWarning:

Parsing dates in DD/MM/YYYY format when dayfirst=False (the default) was specified. This may lead to inconsistently parsed dates! Specify a format to ensure consistent parsing.

<ipython-input-118-a7fea54ccc0d>:2: UserWarning:

Parsing dates in DD/MM/YYYY format when dayfirst=False (the default) was specified. This may lead to inconsistently parsed dates! Specify a format to ensure consistent parsing.



Extracting month and year column from order date

In [ ]:
df1['Month'] = df1['Order Date'].dt.month
df1['Year'] = df1['Order Date'].dt.year

Adding some more columns in our DataFrame 

In [ ]:
df1['Cost'] = df1['Sales']-df1['Profit']

In [ ]:
df1['Cost/Unit'] = df1['Cost']/df1['Quantity']
df1['Price'] = df1['Sales']/df1['Quantity']
df1['Profit/Unit'] = df1['Profit']/df1['Quantity']

In [ ]:
df1.head(1)

Order ID Order Date  Ship Date    Customer Name        Country  \
0  CA-2013-138688 2013-06-13 2013-06-17  Darrin Van Huff  United States   

          City       State Category  \
0  Los Angeles  California   Labels   

                                        Product Name  Sales  Quantity  Profit  \
0  Self-Adhesive Address Labels for Typewriters b...  14.62       2.0    6.87   

   Month  Year  Cost  Cost/Unit  Price  Profit/Unit  
0      6  2013  7.75      3.875   7.31        3.435

setting the columns in our dataframe according to our convenience

In [ ]:
 df = df1.reindex(columns = ['Order ID',	'Order Date',	'Month',	'Year'	,'Ship Date',	'Customer Name',	'Country',	'City',	'State',	'Category',	'Product Name','Quantity','Cost','Sales','Profit','Cost/Unit','Price','Profit/Unit'])

In [ ]:
df.head(2)

Order ID Order Date  Month  Year  Ship Date    Customer Name  \
0  CA-2013-138688 2013-06-13      6  2013 2013-06-17  Darrin Van Huff   
1  CA-2011-115812 2011-09-06      9  2011 2011-06-14  Brosina Hoffman   

         Country         City       State     Category  \
0  United States  Los Angeles  California       Labels   
1  United States  Los Angeles  California  Furnishings   

                                        Product Name  Quantity   Cost  Sales  \
0  Self-Adhesive Address Labels for Typewriters b...       2.0   7.75  14.62   
1  Eldon Expressions Wood and Plastic Desk Access...       7.0  34.69  48.86   

   Profit  Cost/Unit  Price  Profit/Unit  
0    6.87   3.875000   7.31     3.435000  
1   14.17   4.955714   6.98     2.024286

finding no. of years present in our dataframe

In [ ]:
zzz=df['Year'].value_counts().sort_index(ascending=False)
zzz


2014    1099
2013     801
2012     642
2011     661
Name: Year, dtype: int64

**year wise finding-- monthly= (profit and sales)**

In [ ]:
year1=df[df.Year==2011]
year1.shape

(661, 18)

**Now- month wise showing the profit and sales of every state** . 

In [ ]:
q2 = year1.groupby(['Month','State'])['Profit'].sum()
c2 = year1.groupby(['Month','State'])['Sales'].sum()
d2 = pd.merge(q2,c2,how='left',on=['Month','State'])
d2.reset_index(inplace=True)
d2.sort_values(by='Sales', ascending = True)
d2

Month       State   Profit     Sales
0       1     Arizona  -288.01    378.59
1       1  California   380.59   2680.13
2       1    Colorado   252.32   2027.15
3       1      Nevada    19.24     44.79
4       1      Oregon   -33.25    664.76
..    ...         ...      ...       ...
66     12  California  2273.83  20722.05
67     12    Colorado   110.47   1303.21
68     12      Nevada    59.49    475.94
69     12  New Mexico     2.30     57.27
70     12  Washington    32.55    165.45

[71 rows x 4 columns]

In [ ]:
q2 = year1.groupby(['Month','State'])['Profit'].sum()
c2 = year1.groupby(['Month','State'])['Sales'].sum()
d2 = pd.merge(q2,c2,how='left',on=['Month','State'])
d2.reset_index(inplace=True)
d2

Month       State   Profit     Sales
0       1     Arizona  -288.01    378.59
1       1  California   380.59   2680.13
2       1    Colorado   252.32   2027.15
3       1      Nevada    19.24     44.79
4       1      Oregon   -33.25    664.76
..    ...         ...      ...       ...
66     12  California  2273.83  20722.05
67     12    Colorado   110.47   1303.21
68     12      Nevada    59.49    475.94
69     12  New Mexico     2.30     57.27
70     12  Washington    32.55    165.45

[71 rows x 4 columns]

making a line graph by ploty

In [ ]:
#import ploty.express as px
fig = px.line(d2,'Month','Profit',color='State',title='Profit over Months(2011)')
fig.show()

CALIFORNIA profit inc. over month and its highest profit is in month 12, i.e - 2273

In [ ]:
# graph of sales over month
fig = px.line(d2,y='Sales',x='Month',color='State',title = 'Sales_over_month(2011)')
fig.show()

california sales inc. over months and its highest sale is on month 12 which is 20722.

 which state gives you highest profit in the whole year

In [ ]:
state_profit = year1.groupby('State')['Profit'].sum()
state_profit = state_profit.sort_values(ascending=False).reset_index()
state_profit

State    Profit
0  California  12638.00
1  Washington   6607.25
2        Utah    822.67
3    Colorado    438.21
4      Nevada    250.27
5  New Mexico    156.05
6     Montana    129.84
7       Idaho     62.05
8      Oregon   -135.41
9     Arizona   -903.19

In [ ]:
state_profit = year1.groupby('State')['Profit'].sum().reset_index()
fig = px.line(state_profit,x='State',y='Profit',title = 'Yearly profit per state (2011)')
fig.show()

CALIFORNIA is he most profitable state 
and
 ARIZONA is the least profitable state.

 Which state gives most sales to wallmart in 2011 ?

In [ ]:
state_sales = year1.groupby('State')['Sales'].sum()
state_sales = state_sales.sort_values(ascending = False).reset_index()
state_sales

State     Sales
0  California  91303.54
1  Washington  29871.57
2     Arizona   8295.25
3    Colorado   6502.29
4      Oregon   4807.81
5        Utah   2943.69
6      Nevada   2561.20
7  New Mexico    708.02
8       Idaho    465.21
9     Montana    424.49

In [ ]:
state_sales = year1.groupby('State')['Sales'].sum().reset_index()
fig = px.line(state_sales,x='State',y='Sales',title = 'Yearly SALES per State(2011)')
fig.show()

'CALIFORNIA' has highest and 'MONTANA' has lowest SALE in year 2011

In [ ]:
xcz = df.Year.value_counts().sort_index()
#sort_index is used to sort series and sort values is used to sort DataFrame
xcz

2011     661
2012     642
2013     801
2014    1099
Name: Year, dtype: int64

In [ ]:
year2 = df[df['Year']==2012]
year2.shape

(642, 18)

In [ ]:
year2.head()

Order ID Order Date  Month  Year  Ship Date Customer Name  \
13  CA-2012-106320 2012-09-25      9  2012 2012-09-30   Emily Burns   
17  CA-2012-135545 2012-11-24     11  2012 2012-11-30  Kunst Miller   
18  CA-2012-135545 2012-11-24     11  2012 2012-11-30  Kunst Miller   
19  CA-2012-135545 2012-11-24     11  2012 2012-11-30  Kunst Miller   
20  CA-2012-135545 2012-11-24     11  2012 2012-11-30  Kunst Miller   

          Country         City       State     Category  \
13  United States         Orem        Utah       Tables   
17  United States  Los Angeles  California  Accessories   
18  United States  Los Angeles  California      Binders   
19  United States  Los Angeles  California        Paper   
20  United States  Los Angeles  California  Furnishings   

                                         Product Name  Quantity    Cost  \
13      Bretford CR4500 Series Slim Rectangular Table       3.0  804.37   
17  Verbatim 25 GB 6x Blu-ray Single Layer Recorda...       2.0    7.83   
18  Acco PRESSTEX Data Binder with Storage Hooks, ...       6.0   16.46   
19                                         Xerox 1943       3.0   77.77   
20                        Luxo Economy Swing Arm Lamp       4.0   57.43   

      Sales  Profit   Cost/Unit       Price  Profit/Unit  
13  1044.63  240.26  268.123333  348.210000    80.086667  
17    13.98    6.15    3.915000    6.990000     3.075000  
18    25.82    9.36    2.743333    4.303333     1.560000  
19   146.73   68.96   25.923333   48.910000    22.986667  
20    79.76   22.33   14.357500   19.940000     5.582500

Q-1 find profit and sales over months of every state in 2012 

Q-2 find total (profit and sales) over states in year 2012

In [ ]:
a1 = year2.groupby(['Month','State'])['Profit'].sum()
a2 = year2.groupby(['Month','State'])['Sales'].sum()
a3 = pd.merge(a1,a2,how='left',on=['Month','State']).reset_index()
a3

Month       State   Profit     Sales
0       1  California  1016.86  10598.98
1       1    Colorado    17.43    139.42
2       1     Montana    20.81     43.66
3       1      Oregon    51.89    580.68
4       1  Washington   623.03   4279.97
..    ...         ...      ...       ...
69     12    Colorado  -759.83    873.51
70     12     Montana   152.50    487.98
71     12      Nevada    64.82    130.28
72     12      Oregon    36.00    319.97
73     12  Washington   845.46   2955.26

[74 rows x 4 columns]

Q- which state gives most of the profit over time.

In [ ]:
fig = px.line(a3,x='Month',y='Profit',color='State',title='profit over months(2012)')
fig.show()

CALIFORNIA gives most of the profit over time

In [ ]:
fig = px.line(a3,y='Sales',x='Month',color='State',title='Sales over months(2012)')
fig.show()

CALIFORNIA gives max. Sales over time.

Q- Yearly profit of each state.

In [ ]:
state_profit2 = year2.groupby('State')['Profit'].sum()
state_profit2 = state_profit2.sort_values(ascending = False).reset_index()
state_profit2
    


State    Profit
0  California  14371.23
1  Washington   5328.69
2        Utah    669.35
3      Nevada    468.05
4       Idaho    285.32
5     Montana    237.56
6     Arizona    236.40
7  New Mexico    148.73
8      Oregon   -112.08
9    Colorado  -1141.08

In [ ]:
state_profit2 = year2.groupby('State')['Profit'].sum().reset_index()
fig = px.line(state_profit2,x='State',y='Profit',title='Yearly profit per states(2012)')
fig.show()

CALIFORNIA made highest profit of 14371

Q- Yearly Sales of each state

In [ ]:
state_sales2 = year2.groupby('State')['Sales'].sum()
state_sales2 = state_sales2.sort_values(ascending=False).reset_index()
state_sales2

State     Sales
0  California  88443.88
1  Washington  23415.51
2     Arizona   9611.18
3    Colorado   4639.28
4      Oregon   4285.92
5        Utah   3993.32
6      Nevada   1948.46
7       Idaho   1500.24
8  New Mexico   1192.87
9     Montana    935.58

In [ ]:
state_sales2 = year2.groupby('State')['Sales'].sum().reset_index()
fig = px.line(state_sales2,x='State', y='Sales',title='Yearly sales per states(2012)')
fig.show()

CALIFORNIA make a highest sale of 88443

Q-1 find profit and sales over months of every state in 2013

In [ ]:
year3 = df[df['Year']==2013]
year3.shape

(801, 18)

In [ ]:
b1 = year3.groupby(['Month','State'])['Profit'].sum()
b2 = year3.groupby(['Month','State'])['Sales'].sum()
b3 = pd.merge(b1,b2,how='left',on=['Month','State']).reset_index()
b3

Month       State   Profit     Sales
0       1     Arizona    22.38    111.89
1       1  California   666.57   5457.54
2       1  Washington    -8.48    606.61
3       2     Arizona    12.87     61.70
4       2  California   613.21   3793.20
..    ...         ...      ...       ...
62     12  California  2029.92  19086.10
63     12    Colorado    14.37    694.00
64     12  New Mexico     8.23     24.91
65     12      Oregon  -576.09   1759.45
66     12  Washington  1100.00   5224.12

[67 rows x 4 columns]

In [ ]:
fig = px.line(b3,x='Month',y='Profit',color='State',title='Profit over months(2013)')
fig.show()

In [ ]:
fig = px.line(b3,x='Month',y='Sales',color='State',title='Sales over Months(2013)')
fig.show()

**finding Total sales and total profit in year 2013**

In [ ]:
state_profit3 = year3.groupby('State')['Profit'].sum()
state_profit3

State
Arizona       -1485.10
California    19969.58
Colorado      -1389.16
Idaho           280.35
Nevada         2237.69
New Mexico       24.95
Oregon         -565.87
Utah            577.36
Washington     4209.90
Wyoming         100.20
Name: Profit, dtype: float64

In [ ]:
state_profit3 = year3.groupby('State')['Profit'].sum().reset_index()
fig= px.line(state_profit3,x='State',y='Profit',title='Total Profit Over States(2013)')
fig.show()

In [ ]:
state_sales3 = year3.groupby('State')['Sales'].sum()
state_sales3

State
Arizona         6242.05
California    131211.17
Colorado       10666.75
Idaho           1183.49
Nevada          8917.87
New Mexico        64.09
Oregon          5450.84
Utah            1822.41
Washington     19814.28
Wyoming         1603.14
Name: Sales, dtype: float64

In [ ]:
state_sales3 = year3.groupby('State')['Sales'].sum().reset_index()
fig = px.line(state_sales3,x='State',y='Sales',title='total sales over states(2013)')
fig.show()

Now for year 2014

Find Sales and Profit of every states over months

In [ ]:
year4 = df[df['Year']==2014]

In [ ]:
f1 = year4.groupby(['Month','State'])['Profit'].sum()
f2 = year4.groupby(['Month','State'])['Sales'].sum()
f3 = pd.merge(f1,f2,how='left',on=['Month','State']).reset_index()
f3

Month       State   Profit     Sales
0       1     Arizona    11.91    120.04
1       1  California   869.78   4742.61
2       1    Colorado    -0.04    337.68
3       1     Montana  1455.55   4189.38
4       1      Nevada    81.14    239.20
..    ...         ...      ...       ...
72     12  California  3270.77  14211.65
73     12    Colorado  -198.82    197.85
74     12       Idaho     7.99     21.31
75     12  New Mexico    63.12    249.51
76     12  Washington   905.29   4849.75

[77 rows x 4 columns]

In [ ]:
fig = px.line(f3,x='Month',y='Profit',color = 'State',title='Profit Over Months(2014)')
fig.show()

In [ ]:
fig = px.line(f3,x='Month',y='Sales',color = 'State',title='Sales Over Months(2014)')
fig.show()

Q- find total profit and *sales* over states of year 2014

In [ ]:
state_profit4 = year4.groupby('State')['Profit'].sum()
state_profit4 = state_profit4.sort_values(ascending=False)
state_profit4

State
California    29402.79
Washington    17256.86
Montana        1465.92
New Mexico      827.40
Utah            477.18
Nevada          360.75
Idaho           199.01
Oregon         -377.12
Arizona       -1275.98
Colorado      -4435.83
Name: Profit, dtype: float64

In [ ]:
state_profit4 = year4.groupby('State')['Profit'].sum().reset_index()
fig = px.line(state_profit4,x='State',y='Profit',title = 'Total Profit over states(2014)')
fig.show()

In [ ]:
state_sales4 = year4.groupby('State')['Sales'].sum()
state_sales4 = state_sales4.sort_values(ascending = False)
state_sales4


State
California    146729.09
Washington     65539.93
Arizona        11133.54
Colorado       10299.80
Montana         4229.28
Nevada          3301.57
Oregon          2886.57
New Mexico      2818.56
Utah            2460.64
Idaho           1233.55
Name: Sales, dtype: float64

In [ ]:
state_sales4 = year4.groupby('State')['Sales'].sum().reset_index()

In [ ]:
fig = px.line(state_sales4,x='State',y='Sales',title = 'Total Profit over states(2014)')
fig.show()

Overall profit and sales of wallmart store over months

In [ ]:
z1 = df.groupby(['Month','State'])['Profit'].sum()
z2 = df.groupby(['Month','State'])['Sales'].sum()
z3 = pd.merge(z1,z2,how='left',on=['Month','State']).reset_index()
z3

Month       State   Profit     Sales
0        1     Arizona  -253.72    610.52
1        1  California  2933.80  23479.26
2        1    Colorado   269.71   2504.25
3        1     Montana  1476.36   4233.04
4        1      Nevada   100.38    283.99
..     ...         ...      ...       ...
98      12     Montana   152.50    487.98
99      12      Nevada   124.31    606.22
100     12  New Mexico    73.65    331.69
101     12      Oregon  -540.09   2079.42
102     12  Washington  2883.30  13194.58

[103 rows x 4 columns]

In [ ]:
fig = px.line(z3,x='Month',y='Profit',color='State',title = '(Overall) Profit over Months')
fig.show()

In [ ]:
fig = px.line(z3,x='Month',y='Sales',color='State',title = '(Overall) Sales over months')
fig.show()

Total Sales and Profit of Whole wallmart data

In [ ]:
total_state_profit = df.groupby('State')['Profit'].sum()
total_state_profit = total_state_profit.sort_values(ascending= False)
total_state_profit 

State
California    76381.60
Washington    33402.70
Nevada         3316.76
Utah           2546.56
Montana        1833.32
New Mexico     1157.13
Idaho           826.73
Wyoming         100.20
Oregon        -1190.48
Arizona       -3427.87
Colorado      -6527.86
Name: Profit, dtype: float64

In [ ]:
total_state_profit = df.groupby('State')['Profit'].sum().reset_index()
fig = px.line(total_state_profit,x='State', y='Profit',title = '(Overall) profit over states')
fig.show()

In [ ]:
total_state_sales = df.groupby('State')['Profit'].sum()
total_state_sales = total_state_sales.sort_values(ascending= False)
total_state_sales

State
California    76381.60
Washington    33402.70
Nevada         3316.76
Utah           2546.56
Montana        1833.32
New Mexico     1157.13
Idaho           826.73
Wyoming         100.20
Oregon        -1190.48
Arizona       -3427.87
Colorado      -6527.86
Name: Profit, dtype: float64

In [ ]:
total_state_sales = df.groupby('State')['Sales'].sum().reset_index()
fig = px.line(total_state_sales,x='State', y='Sales',title = '(Overall) Sales over States')
fig.show()



Overall conclusion is that - california has highest no. of profit and sales in all over the years and in total years


Now generating category wise sales and profit

In [ ]:
df.head()

Order ID Order Date  Month  Year  Ship Date    Customer Name  \
0  CA-2013-138688 2013-06-13      6  2013 2013-06-17  Darrin Van Huff   
1  CA-2011-115812 2011-09-06      9  2011 2011-06-14  Brosina Hoffman   
2  CA-2011-115812 2011-09-06      9  2011 2011-06-14  Brosina Hoffman   
3  CA-2011-115812 2011-09-06      9  2011 2011-06-14  Brosina Hoffman   
4  CA-2011-115812 2011-09-06      9  2011 2011-06-14  Brosina Hoffman   

         Country         City       State     Category  \
0  United States  Los Angeles  California       Labels   
1  United States  Los Angeles  California  Furnishings   
2  United States  Los Angeles  California          Art   
3  United States  Los Angeles  California       Phones   
4  United States  Los Angeles  California      Binders   

                                        Product Name  Quantity    Cost  \
0  Self-Adhesive Address Labels for Typewriters b...       2.0    7.75   
1  Eldon Expressions Wood and Plastic Desk Access...       7.0   34.69   
2                                         Newell 322       4.0    5.31   
3                     Mitel 5320 IP Phone VoIP phone       4.0  816.43   
4  DXL Angle-View Binders with Locking Rings by S...       3.0   12.72   

    Sales  Profit   Cost/Unit       Price  Profit/Unit  
0   14.62    6.87    3.875000    7.310000     3.435000  
1   48.86   14.17    4.955714    6.980000     2.024286  
2    7.28    1.97    1.327500    1.820000     0.492500  
3  907.15   90.72  204.107500  226.787500    22.680000  
4   18.50    5.78    4.240000    6.166667     1.926667

Q-- How many times Sales occured from each city. 

In [ ]:
cities = df.City.value_counts().reset_index()
cities
cities.rename(columns={'index':'City','City':'Count'})

City  Count
0       Los Angeles    747
1     San Francisco    510
2           Seattle    428
3         San Diego    170
4           Phoenix     63
..              ...    ...
164        Cheyenne      1
165         Yucaipa      1
166         Redding      1
167  Citrus Heights      1
168          Layton      1

[169 rows x 2 columns]

How many times Sales occured from each State

In [ ]:
df.State.value_counts().reset_index()


index  State
0   California   2001
1   Washington    506
2      Arizona    224
3     Colorado    182
4       Oregon    124
5         Utah     53
6       Nevada     39
7   New Mexico     37
8        Idaho     21
9      Montana     15
10     Wyoming      1

How many times Sales occured from each category

In [ ]:
df.Category.value_counts().reset_index()

index  Category
0       Binders       471
1         Paper       450
2   Furnishings       304
3        Phones       277
4       Storage       266
5   Accessories       258
6           Art       250
7        Chairs       207
8    Appliances       136
9        Labels       116
10       Tables       116
11    Bookcases        80
12    Fasteners        72
13     Supplies        69
14    Envelopes        67
15     Machines        39
16      Copiers        25

In [ ]:
z1 = df.groupby('Category')['Profit'].sum().reset_index()
z1=z1.sort_values(by='Profit',ascending=False)
z1

Category    Profit
6       Copiers  19327.25
0   Accessories  16484.62
3       Binders  16096.78
12        Paper  12119.10
13       Phones   9111.06
14      Storage   8645.49
1    Appliances   8261.29
9   Furnishings   7641.26
5        Chairs   4027.62
2           Art   2374.15
10       Labels   2303.07
7     Envelopes   1908.71
16       Tables   1482.54
15     Supplies    626.11
8     Fasteners    275.19
11     Machines   -618.95
4     Bookcases  -1646.50

In [ ]:
z1 = df.groupby('Category')['Profit'].sum().reset_index()
fig = px.line(z1,x='Category',y='Profit',title='Profit over categories')
fig.show()

COPIERS  has the highest profit of 19327

In [ ]:
# Showing above graph by bar graph.
z1 = df.groupby('Category')['Profit'].sum().reset_index()
fig= px.bar(z1,x='Category',y='Profit',color ='Category',title='Profit Over Category',hover_name = 'Category')
fig.show()

In [ ]:
y1 = df.groupby('Category')['Sales'].sum()
y1

Category
Accessories     61114.10
Appliances      30236.32
Art              9212.06
Binders         55961.11
Bookcases       36004.18
Chairs         101781.36
Copiers         49749.23
Envelopes        4118.10
Fasteners         923.20
Furnishings     30072.73
Labels           5078.72
Machines        42444.14
Paper           26663.73
Phones          98684.39
Storage         70532.84
Supplies        18127.12
Tables          84754.60
Name: Sales, dtype: float64

In [ ]:
y1=y1.reset_index()
fig = px.line(y1,x='Category',y='Sales',title = 'Sales over Caegories')
fig.show()

CHAIRS has highest sales of - 101781

In [ ]:
# showing above graph by bar graph
y1=y1.reset_index()
fig = px.bar(y1, x='Category', y='Sales', color='Category', title='Sales Over Category',hover_name= 'Category')
fig.show()

Total profit by state and city in dataframe df

In [ ]:
c = df.groupby(['State','City'])['Profit'].sum()
c

State       City         
Arizona     Avondale            76.17
            Bullhead City        6.72
            Chandler            81.46
            Gilbert            138.25
            Glendale           182.88
                               ...   
Washington  Renton             185.43
            Seattle          29156.13
            Spokane            172.22
            Vancouver           75.46
Wyoming     Cheyenne           100.20
Name: Profit, Length: 170, dtype: float64

Total sales by State and city of DataFrame df

In [ ]:
d = df.groupby(['State','City'])['Sales'].sum()
d

State       City         
Arizona     Avondale            946.81
            Bullhead City        22.29
            Chandler           1076.75
            Gilbert            4172.38
            Glendale           2917.88
                               ...    
Washington  Renton             1242.63
            Seattle          119540.74
            Spokane            2027.92
            Vancouver           686.84
Wyoming     Cheyenne           1603.14
Name: Sales, Length: 170, dtype: float64

In [ ]:
e = df.groupby(['State','City'])['Quantity'].mean()
e
# Q - why we extract mean of this data or mean nikalne se kya faida hota hai...

State       City         
Arizona     Avondale         5.000000
            Bullhead City    1.500000
            Chandler         3.285714
            Gilbert          3.866667
            Glendale         2.913043
                               ...   
Washington  Renton           4.666667
            Seattle          3.714953
            Spokane          5.142857
            Vancouver        3.000000
Wyoming     Cheyenne         4.000000
Name: Quantity, Length: 170, dtype: float64

No. of unique customers present in each city. 

In [ ]:
f=df.groupby(['State','City'])['Customer Name'].nunique()
f

State       City         
Arizona     Avondale           3
            Bullhead City      1
            Chandler           6
            Gilbert            5
            Glendale          13
                            ... 
Washington  Renton             1
            Seattle          188
            Spokane            5
            Vancouver          3
Wyoming     Cheyenne           1
Name: Customer Name, Length: 170, dtype: int64

Name of the customers present in each cities of d/f States 

In [ ]:
w = df.groupby(['State','City'])['Customer Name'].value_counts().reset_index(name='Count').reset_index(drop=True)
w

State           City    Customer Name  Count
0        Arizona       Avondale     Kunst Miller      3
1        Arizona       Avondale     Philip Brown      2
2        Arizona       Avondale          Amy Cox      1
3        Arizona  Bullhead City    Gary Zandusky      2
4        Arizona       Chandler     James Lanier      2
...          ...            ...              ...    ...
1459  Washington        Spokane     Meg O'Connel      1
1460  Washington      Vancouver   Anthony Rawles      3
1461  Washington      Vancouver     Frank Merwin      1
1462  Washington      Vancouver        Jane Waco      1
1463     Wyoming       Cheyenne  Michelle Arnett      1

[1464 rows x 4 columns]

In [ ]:
# now i want to find the no. of orders of each city
s1= df.groupby(['State', 'City']).size() 
s1.name = 'Size'
print(s1)

State       City         
Arizona     Avondale           6
            Bullhead City      2
            Chandler           7
            Gilbert           15
            Glendale          23
                            ... 
Washington  Renton             3
            Seattle          428
            Spokane            7
            Vancouver          5
Wyoming     Cheyenne           1
Name: Size, Length: 170, dtype: int64


In [ ]:
m= pd.merge(c,d,how='left',on=['State','City'])
m=pd.merge(pd.merge(m,e,how='left',on=['State','City']),f,how='left',on=['State','City'])
m= pd.merge(m,s1 ,how='left',on=['State','City'])
m.rename({'Quantity':'Avg. Order Quantity', 'Customer Name_y':'no._of_orders',
'Customer Name_x':'Customers',},axis=1,inplace=True)
m
m.rename(columns={'Quantity':'Avg. Order Quantity'})

Profit      Sales  Avg. Order Quantity  \
State      City                                                      
Arizona    Avondale          76.17     946.81             5.000000   
           Bullhead City      6.72      22.29             1.500000   
           Chandler          81.46    1076.75             3.285714   
           Gilbert          138.25    4172.38             3.866667   
           Glendale         182.88    2917.88             2.913043   
...                            ...        ...                  ...   
Washington Renton           185.43    1242.63             4.666667   
           Seattle        29156.13  119540.74             3.714953   
           Spokane          172.22    2027.92             5.142857   
           Vancouver         75.46     686.84             3.000000   
Wyoming    Cheyenne         100.20    1603.14             4.000000   

                          Customer Name  Size  
State      City                                
Arizona    Avondale                   3     6  
           Bullhead City              1     2  
           Chandler                   6     7  
           Gilbert                    5    15  
           Glendale                  13    23  
...                                 ...   ...  
Washington Renton                     1     3  
           Seattle                  188   428  
           Spokane                    5     7  
           Vancouver                  3     5  
Wyoming    Cheyenne                   1     1  

[170 rows x 5 columns]

In [ ]:
xyz = m.reset_index()
xyz

State           City    Profit      Sales  Avg. Order Quantity  \
0       Arizona       Avondale     76.17     946.81             5.000000   
1       Arizona  Bullhead City      6.72      22.29             1.500000   
2       Arizona       Chandler     81.46    1076.75             3.285714   
3       Arizona        Gilbert    138.25    4172.38             3.866667   
4       Arizona       Glendale    182.88    2917.88             2.913043   
..          ...            ...       ...        ...                  ...   
165  Washington         Renton    185.43    1242.63             4.666667   
166  Washington        Seattle  29156.13  119540.74             3.714953   
167  Washington        Spokane    172.22    2027.92             5.142857   
168  Washington      Vancouver     75.46     686.84             3.000000   
169     Wyoming       Cheyenne    100.20    1603.14             4.000000   

     Customer Name  Size  
0                3     6  
1                1     2  
2                6     7  
3                5    15  
4               13    23  
..             ...   ...  
165              1     3  
166            188   428  
167              5     7  
168              3     5  
169              1     1  

[170 rows x 7 columns]

Q- Is there any relation between Sales & Profit

In [ ]:
 a = df.groupby(['State','City'])['Profit'].sum()

In [ ]:
b=df.groupby(['State','City'])['Sales'].sum()

In [ ]:
ab = pd.merge(a,b,how='left',on=['State','City']).reset_index()
ab.sort_values(by = 'Profit').reset_index(drop=True)
ab

State           City    Profit      Sales
0       Arizona       Avondale     76.17     946.81
1       Arizona  Bullhead City      6.72      22.29
2       Arizona       Chandler     81.46    1076.75
3       Arizona        Gilbert    138.25    4172.38
4       Arizona       Glendale    182.88    2917.88
..          ...            ...       ...        ...
165  Washington         Renton    185.43    1242.63
166  Washington        Seattle  29156.13  119540.74
167  Washington        Spokane    172.22    2027.92
168  Washington      Vancouver     75.46     686.84
169     Wyoming       Cheyenne    100.20    1603.14

[170 rows x 4 columns]

**Graph of Sales and Profit Co-Relation by State**

In [ ]:
fig = px.scatter(ab,x='Sales',y='Profit',color='State',title='Sales and Profit Co-Relation by State',size='Sales',hover_name='State')
fig.show()

Graph of Sales & Profit Co-Relation by City

In [ ]:
fig=px.scatter(ab,'Sales','Profit',color='State',title='Sales & Profit Co-Relation by City',
               size='Sales', hover_name='City')
fig.show()

 Q--Is there any relation between No. of Customers & Sales?

In [ ]:
zx = df.groupby(['State'])['Customer Name'].nunique().reset_index()   
zx.rename(columns={'Customer Name':'No. Of Customers'},inplace=True)
zy = df.groupby(['State'])['Sales'].sum()
zxy = pd.merge(zx,zy,how='left',on=['State']).reset_index()
zxy

index       State  No. Of Customers      Sales
0       0     Arizona               100   35282.02
1       1  California               577  457687.68
2       2    Colorado                75   32108.12
3       3       Idaho                11    4382.49
4       4     Montana                 8    5589.35
5       5      Nevada                23   16729.10
6       6  New Mexico                22    4783.54
7       7      Oregon                51   17431.14
8       8        Utah                26   11220.06
9       9  Washington               224  138641.29
10     10     Wyoming                 1    1603.14

In [ ]:
fig=px.scatter(zxy,'No. Of Customers','Sales',color='State',title='No. of Customers & Sales Co-Relation',
               hover_name='State',size='No. Of Customers')
fig.show()

Q-- What is Total share of States in genrating profit?

In [ ]:
wx = df.groupby(['State'])['Profit'].sum().reset_index()
wx

State    Profit
0      Arizona  -3427.87
1   California  76381.60
2     Colorado  -6527.86
3        Idaho    826.73
4      Montana   1833.32
5       Nevada   3316.76
6   New Mexico   1157.13
7       Oregon  -1190.48
8         Utah   2546.56
9   Washington  33402.70
10     Wyoming    100.20

In [ ]:
fig= px.pie(wx,names=wx['State'],values=wx['Profit'],title='Profit Output By State')
fig.show()

Q--. What is the share of each states in genrating sales for wallmart?

In [ ]:
 wy = df.groupby('State')['Sales'].sum().reset_index()
 wy

State      Sales
0      Arizona   35282.02
1   California  457687.68
2     Colorado   32108.12
3        Idaho    4382.49
4      Montana    5589.35
5       Nevada   16729.10
6   New Mexico    4783.54
7       Oregon   17431.14
8         Utah   11220.06
9   Washington  138641.29
10     Wyoming    1603.14

In [ ]:
fig = px.pie(wy,'State','Sales',title='Sales output by each state',hover_name='State')
fig.show()

California has the biggest share and genrates 63% of wallmart's revennue. Washington stand 2nd with contribution of 19.1%.

Q--. What is the share of each state in terms of footfall?    ...# footfall means how many customers  came into wallmart

In [ ]:
wz = df.groupby('State')['Customer Name'].nunique().reset_index()
wz.rename(columns={'Customer Name':'No. of unique Customers'},inplace=True)
wz

State  No. of unique Customers
0      Arizona                      100
1   California                      577
2     Colorado                       75
3        Idaho                       11
4      Montana                        8
5       Nevada                       23
6   New Mexico                       22
7       Oregon                       51
8         Utah                       26
9   Washington                      224
10     Wyoming                        1

In [ ]:
fig = px.pie(wz,'State','No. of unique Customers',title= 'footfall per State',hover_name='State')
fig.show()

California is the 1st in the terms of footfall and has 61.5% of total footfall in wallmart.

Q--.In each State which is the best performing city in term of no. of customers?

In [ ]:
xs = df.groupby(['State','City'])['Customer Name'].nunique().reset_index()
es = xs.groupby(['State'])['Customer Name'].idxmax().reset_index(drop=True)
result = xs.loc[es]
result

State         City  Customer Name
7       Arizona      Phoenix             28
46   California  Los Angeles            304
99     Colorado       Denver             17
110       Idaho        Boise              3
118     Montana  Great Falls              3
122      Nevada    Las Vegas              8
126  New Mexico  Albuquerque              7
137      Oregon     Portland             15
147        Utah         Orem              5
166  Washington      Seattle            188
169     Wyoming     Cheyenne              1

In [ ]:
fig = px.bar(result,'State','Customer Name',color='State',title='"Best Performing Cities by Number of Customers per State"')
fig.show()

In [ ]:
# now find top 10 customers which generated highest profit

In [ ]:
cv=df.groupby(['State','Customer Name'])['Profit'].sum()
yv=df.groupby(['State','Customer Name'])['Quantity'].sum()
iv=df.groupby(['State','Customer Name'])['Sales'].sum()
qv=pd.merge(pd.merge(cv,yv,how='left',on=['State','Customer Name']),iv,how='left',on=['State','Customer Name'])
qv

Profit  Quantity    Sales
State      Customer Name                               
Arizona    Adrian Barton     -186.93      13.0   985.47
           Adrian Hane        -25.91       4.0   259.14
           Alex Grayson      -135.09       2.0   182.55
           Allen Armold         1.47       2.0    16.77
           Amy Cox             83.51       8.0   742.34
...                              ...       ...      ...
Washington Victoria Wilson      1.53       3.0     5.28
           Vivek Gonzalez      25.51       2.0   115.96
           Yana Sorensen     1255.25      19.0  3859.49
           Zuschuss Carroll   379.50      15.0  1549.77
Wyoming    Michelle Arnett    100.20       4.0  1603.14

[1118 rows x 3 columns]

In [ ]:
cv = df.groupby(['State','Customer Name'])['Profit'].sum().reset_index()
cv = cv.sort_values(by='Profit',ascending=False).reset_index().head(10)
cv

index       State    Customer Name   Profit
0   1066  Washington     Raymond Buch  6734.47
1    350  California        Jane Waco  2069.63
2    972  Washington     Fred Hopkins  1930.47
3    774      Nevada    Dennis Pardue  1691.31
4    573  California    Robert Marley  1598.51
5    763     Montana   Alan Dominguez  1455.55
6   1115  Washington    Yana Sorensen  1255.25
7   1005  Washington      Joseph Holt  1076.91
8    973  Washington       Gary Hwang  1073.14
9    548  California  Penelope Sewall  1056.31

In [ ]:
xx = qv.nlargest(10,'Profit')
xx.reset_index(inplace=True) # or assign reset_index() to a variable
xx

State    Customer Name   Profit  Quantity     Sales
0  Washington     Raymond Buch  6734.47       9.0  14052.48
1  California        Jane Waco  2069.63      33.0   7381.89
2  Washington     Fred Hopkins  1930.47      12.0   4607.27
3      Nevada    Dennis Pardue  1691.31      12.0   4685.08
4  California    Robert Marley  1598.51      17.0   4881.13
5     Montana   Alan Dominguez  1455.55      13.0   4189.38
6  Washington    Yana Sorensen  1255.25      19.0   3859.49
7  Washington      Joseph Holt  1076.91      23.0   3115.88
8  Washington       Gary Hwang  1073.14      11.0   3290.52
9  California  Penelope Sewall  1056.31       8.0   3372.85

 Q--What 10 Customers have genrated us the most Profit?

In [ ]:
fig=px.bar(xx,'Customer Name','Profit',color='Customer Name',title= 'Top 10 most profitable customer')
fig.show()

Q--In each state which is the top customer in terms of profit?

In [ ]:
top_cs = df.groupby(['State','Customer Name'])['Profit'].sum().reset_index()
max_idx=top_cs.groupby(['State'])['Profit'].idxmax()
max_value=top_cs.loc[max_idx].reset_index(drop=True)
max_value

State      Customer Name   Profit
0      Arizona        John Murray   319.17
1   California          Jane Waco  2069.63
2     Colorado  Russell Applegate   252.32
3        Idaho   Justin Deggeller   266.17
4      Montana     Alan Dominguez  1455.55
5       Nevada      Dennis Pardue  1691.31
6   New Mexico    Benjamin Farhat   322.94
7       Oregon        Clay Ludtke   156.30
8         Utah  Valerie Dominguez   576.56
9   Washington       Raymond Buch  6734.47
10     Wyoming    Michelle Arnett   100.20

Q--In each state which is the top customer in terms of profit?

In [ ]:
fig= px.bar(max_value,'State','Profit',color='State',hover_name='Customer Name')
fig.show()

Q--. Which Top 10 customers has caused the most losses to us?

In [ ]:
most_losses = df.groupby(['State','City','Customer Name'])['Profit'].sum().reset_index()
most_losses = most_losses.sort_values(by='Profit', ascending=True).head(10).reset_index(drop=True)
most_losses = most_losses.rename(columns={'Profit':'Loss'})
most_losses

State              City     Customer Name     Loss
0  Colorado        Louisville    Sharelle Roach -3424.35
1  Colorado            Pueblo    Laurel Beltran  -961.53
2   Arizona           Phoenix     Tanja Norvell  -857.19
3  Colorado  Colorado Springs       Jim Radford  -785.16
4   Arizona           Phoenix      Resi Pölking  -694.29
5   Arizona           Phoenix      Maris LaWare  -482.59
6   Arizona              Yuma       Gary Hansen  -479.99
7  Colorado            Aurora       Noah Childs  -461.20
8   Arizona           Phoenix  Maureen Fritzler  -459.99
9   Arizona           Phoenix    Laurel Workman  -393.34

In [ ]:
fig=px.bar(most_losses,'Customer Name','Loss',color='Customer Name',title= 'Top 10 losses in Wallmart',hover_name='City')
fig.show()

Q-- In each state which is the Top customer in terms of losess

In [ ]:
dp = df.groupby(['State','City','Customer Name'])['Profit'].sum().reset_index()

sp=dp.sort_values(by='Profit',ascending = True)
pdd = sp.groupby('State')['Profit'].idxmin()
result_2 = sp.loc[pdd].reset_index(drop=True)
result_2= result_2.rename(columns={'Profit':'Loss'})
result_2
# as we use idxmax- to get index of max value, - and -same we can use idxmin- to get index of min value 

State           City      Customer Name     Loss
0      Arizona        Phoenix      Tanja Norvell  -857.19
1   California  San Francisco  Charles McCrossin  -325.57
2     Colorado     Louisville     Sharelle Roach -3424.35
3        Idaho          Boise       Noel Staavos     1.12
4      Montana    Great Falls       Eugene Moren     2.21
5       Nevada           Reno   Neil Französisch     0.00
6   New Mexico     Farmington       Anthony Witt     0.39
7       Oregon    Springfield      Juliana Krohn  -386.22
8         Utah           Lehi         Karen Seio     1.68
9   Washington        Seattle       Tim Brockman   -52.07
10     Wyoming       Cheyenne    Michelle Arnett   100.20

In [ ]:
fig = px.bar(result_2,'State','Loss',color='State',hover_name='Customer Name')
fig.show()

Monthly Sales over months of all years ?

In [ ]:
monthly_sales = df.groupby(pd.Grouper(key='Order Date', freq='M'))['Sales'].sum()
monthly_sales


Order Date
2011-01-31     6272.91
2011-02-28     2034.87
2011-03-31     9611.27
2011-04-30     9279.30
2011-05-31     6646.20
2011-06-30     8272.06
2011-07-31    21058.14
2011-08-31    10987.62
2011-09-30    20667.71
2011-10-31    11996.40
2011-11-30    16144.08
2011-12-31    24912.51
2012-01-31    15642.71
2012-02-29     6634.08
2012-03-31    16783.35
2012-04-30     8270.37
2012-05-31    12939.23
2012-06-30     7959.27
2012-07-31     7717.10
2012-08-31    11409.87
2012-09-30    10696.94
2012-10-31    13050.03
2012-11-30    12089.36
2012-12-31    16773.93
2013-01-31     6176.04
2013-02-28     5517.33
2013-03-31    14190.44
2013-04-30    14073.33
2013-05-31    16665.03
2013-06-30    15586.66
2013-07-31    14701.79
2013-08-31    16563.33
2013-09-30    17957.34
2013-10-31    11799.99
2013-11-30    25952.08
2013-12-31    27792.73
2014-01-31    11935.46
2014-02-28    18510.72
2014-03-31    36484.24
2014-04-30    22848.25
2014-05-31    13228.69
2014-06-30    13696.78
2014-07-31    16780.04
